keras_nlp/metrics/perplexity.py

```diff
@@ -0,0 +1,166 @@
+    >>> target = tf.random.uniform(
+    ...     shape=[2, 5],  maxval=10, dtype=tf.int32, seed=42)
+    >>> logits = tf.random.uniform(shape=(2, 5, 10), seed=42)
+    >>> sample_weight = tf.cast(
+    ...     tf.math.logical_not(tf.equal(target, 0)), tf.float32)
+    >>> perplexity.update_state(target, logits, sample_weight)
+    >>> perplexity.result()
+    <tf.Tensor: shape=(), dtype=float32, numpy=13.1128>
+
+    2. Call perplexity directly.
+    >>> tf.random.set_seed(42)
+    >>> perplexity = keras_nlp.metrics.Perplexity(name="perplexity")
+    >>> target = tf.random.uniform(
+    ...     shape=[2, 5],  maxval=10, dtype=tf.int32, seed=42)
+    >>> logits = tf.random.uniform(shape=(2, 5, 10), seed=42)
+    >>> perplexity(target, logits)
+    <tf.Tensor: shape=(), dtype=float32, numpy=11.8781595>
+
+    3. Provide the padding token ID and let the class compute the mask on its
+       own.
+    >>> tf.random.set_seed(42)
+    >>> perplexity = keras_nlp.metrics.Perplexity(
+    ...     name="perplexity", pad_token_id=0)
+    >>> target = tf.random.uniform(
+    ...     shape=[2, 5],  maxval=10, dtype=tf.int32, seed=42)
+    >>> logits = tf.random.uniform(shape=(2, 5, 10), seed=42)
+    >>> perplexity(target, logits)
+    <tf.Tensor: shape=(), dtype=float32, numpy=13.1128>
+    """
+
+    def __init__(
+        self,
+        name="perplexity",
+        dtype=None,
+        from_logits=False,
+        pad_token_id=None,
+        **kwargs,
+    ):
+        super().__init__(name=name, dtype=dtype, **kwargs)
+
+        if not tf.as_dtype(self.dtype).is_floating:
+            raise ValueError(
+                "`dtype` must be a floating point type. "
+                f"Received: dtype={dtype}"
+            )
+
+        self._cross_entropy = keras.losses.SparseCategoricalCrossentropy(
+            from_logits=from_logits, reduction="sum"
+        )
+
+        self.pad_token_id = pad_token_id
+
+        self._aggregate_cross_entropy = self.add_weight(
+            name="aggregate_cross_entropy",
+            initializer="zeros",
+            dtype=self.dtype,
+        )
+        self._number_of_samples = self.add_weight(
+            name="number_of_samples", initializer="zeros", dtype=self.dtype
+        )
+
+    def update_state(self, y_true, y_pred, sample_weight=None):
+        # y_true shape: (batch_size, seq_len)
+        # y_pred shape: (batch_size, seq_len, vocab_size)
+        y_true = tf.cast(y_true, self.dtype)
+        y_pred = tf.cast(y_pred, self.dtype)
+        batch_size = tf.cast(tf.shape(y_true)[0], self.dtype)
+
+        if self.pad_token_id is not None:
+            sample_weight = tf.cast(
+                tf.math.logical_not(tf.equal(y_true, self.pad_token_id)),
+                self.dtype,
+            )
+
+        if sample_weight is not None:
+            sample_weight = tf.cast(sample_weight, self.dtype)
+
+        # Calculate the Cross Entropy Loss.
+        cross_entropy_value = tf.cast(
```

Comment:
I checked out the source code of tf.keras.metrics.SparseCategoricalCrossentropy, and it is dong WEIGHTED_MEAN reduction (https://github.com/keras-team/keras/blob/d8fcb9d4d4dad45080ecfdd575483653028f8eda/keras/metrics.py#L583), which should automatically set the divisor as the sum over masks, could you help verify it with your unit test? Thanks!

Reply:
Sure. Will try this out! Thanks!

Reply:
@chenmoneygithub, this particular UT failed:

```
____________________________________________________ PerplexityTest.test_two_inputs_from_logits ____________________________________________________ 

self = <keras_nlp.metrics.perplexity_test.PerplexityTest testMethod=test_two_inputs_from_logits>

    def test_two_inputs_from_logits(self):
        perplexity = Perplexity(from_logits=True, pad_token_id=0)

        y_true_1 = tf.constant([[1, 3, 0], [2, 1, 3]])
        y_pred_1 = tf.constant(
            [
                [
                    [1.034, 4.797, 2.82, 1.154],
                    [2.258, 1.591, 1.811, 1.852],
                    [3.216, 1.037, 0.3662, 2.7],
                ],
                [
                    [1.363, 1.726, 1.898, 2.582],
                    [1.163, 1.943, 1.761, 1.497],
                    [2.766, 1.453, 2.61, 2.805],
                ],
            ]
        )

        perplexity_val = perplexity(y_true_1, y_pred_1)
        self.assertAlmostEqual(perplexity_val, 2.8788896)

        y_true_2 = tf.constant([[2, 0, 0], [1, 2, 3]])
        y_pred_2 = tf.constant(
            [
                [
                    [2.887, 0.885, 2.973, 2.582],
                    [0.3838, 2.629, 1.91, 1.802],
                    [0.2578, 1.081, 1.125, 2.773],
                ],
                [
                    [1.623, 2.784, 0.2109, 2.66],
                    [2.395, 2.01, 0.252, 1.828],
                    [0.4482, 2.629, 0.9697, 0.998],
                ],
            ]
        )
        perplexity_val = perplexity(y_true_2, y_pred_2)
>       self.assertEqual(perplexity_val, 3.9998498)
E       AssertionError: <tf.Tensor: shape=(), dtype=float32, numpy=3.3319612> != 3.9998498

keras_nlp\metrics\perplexity_test.py:132: AssertionError
```

Reply:
I did a quick analysis on Colab. Apparently, when `sample_weight` is not provided, the outputs of `keras.losses.SparseCategoricalCrossentropy` and `keras.metrics.SparseCategoricalCrossentropy` don't match. Have a look at this: https://colab.research.google.com/drive/1Jh44hylKVmmdqR1Z3B_redGzr1ZJdI3i?usp=sharing.

Let me know what the correct course of action is. Thanks!

Reply:
This is pretty odd, I guess we can stick to Loss function and open an issue for future investigation.

Reply:
Coolio :). Thanks! 👍🏼